### PR TITLE
Detect unsatisfiable range queries; warn and short-circuit.

### PR DESCRIPTION
**Pull request by gefjon**
# Description of Changes

This commit fixes a panic caused by unsatisfiable range bounds on an index query, e.g. `WHERE x < 5 AND x > 5`. These unsatisfiable bounds made Rust's `BTreeMap` angry (See https://doc.rust-lang.org/src/alloc/collections/btree/search.rs.html#106-124), and panicked. They also represent probable bugs, as it's silly to write a query which statically will return no rows.

With this commit, we detect statically unsatisfiable bounds in two cases:
- When compiling queries, we log a message at `WARN` containing the offending query.
- When evaluating queries, we silently construct an `EmptyRelOps` rather than a real query iterator.

This commit also adds a test that the offending queries can be compiled and executed without panicking, and select no rows.

# API and ABI breaking changes

N/a

If this is an API or ABI breaking change, please apply the
corresponding GitHub label.

# Expected complexity level and risk

1

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

# Testing

*Describe any testing you've done, and any testing you'd like your reviewers to do,
so that you're confident that all the changes work as expected!*

- [x] Checked that the logs are emitted by enabling `env_logger` and passing `--nocapture` when running the tests.
- [x] Newly-added test `test_impossible_bounds_no_panic` used to panic but doesn't anymore.

### CODE REVIEW

File: crates/vm/src/expr.rs [hunk 1]
```diff
@@ -1436,14 +1446,24 @@ impl QueryExpr {
                 self.query.push(Query::IndexScan(IndexScan { table, columns, bounds }));
                 self
             }
-            // merge with a preceding lower bounded index scan (inclusive)
+            // merge with a preceding lower bounded index scan (enclusive)
             Query::IndexScan(IndexScan {
```
Comment (mamcx) [inline crates/vm/src/expr.rs:1450]: This word [exist](https://wikidiff.com/enclusive/inclusive)?

**Review by kazimuth — APPROVED** [review]
LGTM. 

Eventually we might want a system for lint passes but I think the places this happens right now make sense.

File: crates/vm/src/expr.rs [hunk 2]
```diff
@@ -1356,8 +1356,18 @@ impl QueryExpr {
                 bounds: (Bound::Unbounded, Bound::Excluded(upper)),
                 ..
             }) if columns == lhs_col_id => {
+                // Queries like `WHERE x < 5 AND x > 5` never return any rows and are likely mistakes.
+                // Detect such queries and log a warning.
+                // Compute this condition early, then compute the resulting query and log it.
+                let is_never = !inclusive && value == upper;
```
Comment (joshua-spacetime) [inline crates/vm/src/expr.rs:1362]: We should not be creating an index scan plan in this case.

**Review by joshua-spacetime — COMMENTED** [review]
If you'd like to merge this, that's fine, I'd just like comments around the log lines and the fix that say this is a temporary fix and that we should not be generating invalid index scan plans.

**Comment by gefjon** [comment]
> If you'd like to merge this, that's fine, I'd just like comments around the log lines and the fix that say this is a temporary fix and that we should not be generating invalid index scan plans.

Agreed. I've added comments to that effect. I'll hold off on merging into master for now, though we may include this in the build used for the BitCraft alpha.

File: crates/vm/src/expr.rs (hunk 1 above)
Comment (gefjon) [reply crates/vm/src/expr.rs:1456]: 🤦. I replaced one letter but not the other when fixing a copy-paste error. Pushed a fix.

**Pull request merged**